File: research/experiments/sweep/runs/mathexpr-full-csharp-sonnet45/MathExprParser/Evaluator.cs
namespace MathExprParser;

public static class Evaluator
{
    public static double Evaluate(AstNode node)
    {
        return node switch
        {
            NumberLiteral num => num.Value,
            UnaryExpr unary => EvaluateUnary(unary),
            BinaryExpr binary => EvaluateBinary(binary),
            _ => throw new Exception($"Unknown AST node type: {node.Type}")
        };
    }

    private static double EvaluateUnary(UnaryExpr unary)
    {
        var operandValue = Evaluate(unary.Operand);
        return unary.Op switch
        {
            "-" => -operandValue,
            _ => throw new Exception($"Unknown unary operator: {unary.Op}")
        };
    }

    private static double EvaluateBinary(BinaryExpr binary)
    {
        var leftValue = Evaluate(binary.Left);
        var rightValue = Evaluate(binary.Right);

        return binary.Op switch
        {
            "+" => leftValue + rightValue,
            "-" => leftValue - rightValue,
            "*" => leftValue * rightValue,
            "/" => rightValue == 0 ? throw new Exception("Division by zero") : leftValue / rightValue,
            "%" => rightValue == 0 ? throw new Exception("Modulo by zero") : leftValue % rightValue,
            "**" => Math.Pow(leftValue, rightValue),
            _ => throw new Exception($"Unknown binary operator: {binary.Op}")
        };
    }
}

File: research/experiments/sweep/runs/mathexpr-full-kotlin-sonnet45/src/main/kotlin/mathexpr/Evaluator.kt
package mathexpr

import kotlin.math.pow

fun evaluate(node: AstNode): Double {
    return when (node) {
        is AstNode.NumberLiteral -> node.value
        is AstNode.UnaryExpr -> {
            val operandValue = evaluate(node.operand)
            when (node.op) {
                "-" -> -operandValue
                else -> throw IllegalArgumentException("Unknown unary operator: ${node.op}")
            }
        }
        is AstNode.BinaryExpr -> {
            val leftValue = evaluate(node.left)
            val rightValue = evaluate(node.right)
            when (node.op) {
                "+" -> leftValue + rightValue
                "-" -> leftValue - rightValue
                "*" -> leftValue * rightValue
                "/" -> {
                    if (rightValue == 0.0) {
                        throw ArithmeticException("Division by zero")
                    }
                    leftValue / rightValue
                }
                "%" -> {
                    if (rightValue == 0.0) {
                        throw ArithmeticException("Modulo by zero")
                    }
                    leftValue % rightValue
                }
                "**" -> leftValue.pow(rightValue)
                else -> throw IllegalArgumentException("Unknown binary operator: ${node.op}")
            }
        }
    }
}
